src/backend/replication/logical/message.c

```diff
@@ -93,3 +96,82 @@ logicalmsg_redo(XLogReaderState *record)
 
 	/* This is only interesting for logical decoding, see decode.c. */
 }
+
+/*
+ * NEON: remove AUX object
+ */
+void
+wallog_file_removal(char const* path)
+{
+	char	prefix[MAXPGPATH];
+
+	/* Do not wallog AUX file at replica */
+	if (!XLogInsertAllowed())
+		return;
+
+	snprintf(prefix, sizeof(prefix), "neon-file:%s", path);
+	elog(DEBUG1, "neon: deleting contents of file %s", path);
+
+	/* unlink file */
+	LogLogicalMessage(prefix, NULL, 0, false, true);
+}
+
+/*
+ * NEON: persist file in WAL to save it in persistent storage.
+ * This function changes current position in the file, so caller should be aware of it.
```

Comment:
Changing the current file position still scares me. I guess it works, but it's scary. If you look at the caller in logical_heap_rewrite_flush_mappings() for example, it depends on the fact that the position is restored to the end of the file after the operation. And/or the fact the the File*() functions use pread/pwrite and don't depend on the fd position.

It would be pretty straightforward to not change the file position here. Just use fstat() to get the file size and pread() for the reads. I suggest we do that to remove this footgun.

Reply:
Fixed